### ISSUE TITLE
Using quantized tflite models

### ISSUE THREAD

**Issue by bjoernholzhauer** [shaqian/flutter_tflite#55]
When I substitute a quantized model into code that works for image classification with the non-quantized model (I simply substituted 'mobilenet_v2_1.0_224_quant.tflite' for 'Mobilenet_V2_1.0_224'), I get:
`Caused by: java.lang.IllegalArgumentException: Cannot convert between a TensorFlowLite tensor with type UINT8 and a Java object of type [[F (which is compatible with the TensorFlowLite type FLOAT32).`

Is it possible to use quantized models? If so, how (would be good to have something in the documentation)? If not, would also be good if the documentation said so - or if it is just not possible at the moment, but hopefully in the future, I guess consider this a feature request.

**Comment by securingsincity** [comment]
Based on some searching of issues last night #53 and #59 both are related to this issue. The automl vision edge outputs a quantized tflite model.

 Here are two images from netron describing the differences between the quantized model and the mobilenet v2 model that flutter_tflite currently supports

<img width="493" alt="Screen Shot 2019-10-19 at 9 44 55 PM" src="https://user-images.githubusercontent.com/1388049/67161325-624d0680-f327-11e9-811e-ca3f3edb9daa.png">
<img width="506" alt="Screen Shot 2019-10-19 at 9 44 37 PM" src="https://user-images.githubusercontent.com/1388049/67161326-624d0680-f327-11e9-8f27-4aed9f4c46cc.png">

Note that they are the same except one accepts a uint8 list and the other takes a float32 list. 

I'm not entirely sure what would need to change on the flutter tflite side to support this kind of model but hopefully this helps

**Comment by Statyk7** [comment]
I'm having the same problem and wondering if there is a way or workaround to handle those models?

**Comment by waltermaldonado** [comment]
For what I have seen, if you use the method to run the detections on binary you can use a quantized model. Actually, the conversion from image to ByteList suggested in the docs is made considering 8-bit integer as unit, as you can see below:

```
Uint8List imageToByteListUint8(img.Image image, int inputSize) {
  var convertedBytes = Uint8List(1 * inputSize * inputSize * 3);
  var buffer = Uint8List.view(convertedBytes.buffer);
  int pixelIndex = 0;
  for (var i = 0; i < inputSize; i++) {
    for (var j = 0; j < inputSize; j++) {
      var pixel = image.getPixel(j, i);
      buffer[pixelIndex++] = img.getRed(pixel);
      buffer[pixelIndex++] = img.getGreen(pixel);
      buffer[pixelIndex++] = img.getBlue(pixel);
    }
  }
  return convertedBytes.buffer.asUint8List();
}
```
This conversion should work for a quantized model, but is not working for a non-quantized one. The convertedBytes List should be 4 times the one is being suggested to work for non-quantized models.

When I use the detections on image path it works perfectly.

Please correct me if I am wrong, but it does not look like it is possible to use binary detections on a non-quantized model, since it takes only UInt8List as input param (I really hope I am wrong).

**Comment by Statyk7** [comment]
Have you been able to run the MobileNet quantized version?
Can be found here: https://www.tensorflow.org/lite/guide/hosted_models
I have no success with Mobilenet_V1_1.0_224_quant :(
I have tried with runModelOnImage and with runModelOnBinary using the image to byte functions... no results... (and no errors)

But when using the TensorFlow iOS Sample App it works just fine!
https://github.com/tensorflow/examples/tree/master/lite/examples/image_classification/ios

**Comment by waltermaldonado** [comment]
No, I've never tried those models, but I think them should work aswell. Let us see your code, maybe we can find something...

**Comment by Statyk7** [comment]
I'm using the example provided with the tflite package:
https://github.com/shaqian/flutter_tflite/tree/master/example

With an additional asset for the model (the labels are the same than for the non-quantized model) in pubspec.yaml:
`   - assets/mobilenet_v1_1.0_224_quant.tflite
`
Then I load the quantized model instead of the non-quantized one in main.dart:loadModel:
default:
`          res = await Tflite.loadModel(
            model: "assets/mobilenet_v1_1.0_224_quant.tflite",
            labels: "assets/mobilenet_v1_1.0_224.txt",
`
That's it!

**Comment by waltermaldonado** [comment]
Just to clarify, is your non-quantized model a *detection* model (localization + classification)? Because it seems to me that those quantized models are *classification* only models.

**Comment by Statyk7** [comment]
It's an image classification model I believe...

**Comment by Ehtasha** [comment]
@Statyk7  @waltermaldonado  

I'm integrating my own custom model in this example but the app crashes when we send an image to the model using method segmentMobileNet.

I have also tried with [runModelOnBinary] but issue still stand.

My custom model is trained on PyTorch and I have converted into Tensorflow using onnx and then in .tflite.
Model is not quantized.

**Comment by yumemi-RyoShimizu** [comment]
I created an image labeling model with AutoML, and since the model should have been quantized, I converted the image to uint8, but the following error was output.
`Caused by: java.lang.IllegalArgumentException: Cannot convert between a TensorFlowLite tensor with type UINT8 and a Java object of type [(F (which is compatible with the TensorFlowLite type FLOAT32) .`

**Comment by AndrsDev** [comment]
I'm having the same problem!!! is there any updates of this?

**Comment by AndrsDev** [comment]
**Here are my Auto ML properties**

Throws error 
`
Caused by: java.lang.IllegalArgumentException: Cannot convert between a TensorFlowLite tensor with type UINT8 and a Java object of type [(F (which is compatible with the TensorFlowLite type FLOAT32) .
`

<img width="497" alt="Screen Shot 2020-02-28 at 10 55 18 AM" src="https://user-images.githubusercontent.com/34043452/75568394-d0fa4900-5a18-11ea-9064-e26bc2d896ec.png">

**Comment by oncul** [comment]
Do u have problem with AutoMl generated tflite file on ios?

**Comment by L-is-0** [comment]
@AndrsDev I have the same error here

**Comment by joknjokn** [comment]
Did anyone find a solution to this? 

Also getting:
`java.lang.IllegalArgumentException: Cannot convert between a TensorFlowLite tensor with type UINT8 and a Java object of type [(F (which is compatible with the TensorFlowLite type FLOAT32)`

I'm trying with this model, and livestreamed camera image (YUV on android):
https://tfhub.dev/google/lite-model/aiy/vision/classifier/birds_V1/2

The page states:

> Inputs are expected to be 3-channel RGB color images of size 224 x 224, scaled to [0, 1].
> This model outputs to image_classifier.

I've tried a million things now and I can't get it to work. If I try to convert the streamed image to RGB, I get the UINT8/FLOAT32-error above.

**Comment by Bryanx** [comment]
I solved this by using [tflite_flutter](https://github.com/am15h/tflite_flutter_plugin) and tflite_flutter_helper instead of this library. Here is a gist here in case anyone is running into this as well: https://gist.github.com/Bryanx/b839e3ceea0f9647ffbc5f90e3091742.

**Comment by tobiascornille** [comment]
@Bryanx Do you think `tflite_flutter_helper ` alone would solve the issue? I.e. is it compatible with this library?

**Comment by zoraiz-WOL** [comment]
use this code to train you custom model


import os

import numpy as np

import tensorflow as tf
assert tf.__version__.startswith('2')

from tflite_model_maker import model_spec
from tflite_model_maker import image_classifier
from tflite_model_maker.config import ExportFormat
from tflite_model_maker.config import QuantizationConfig
from tflite_model_maker.image_classifier import DataLoader

import matplotlib.pyplot as plt

#to unzip a rar
!unzip path-of-zip-file -d path-to-save-extract-file



data = DataLoader.from_folder('path-of-custom-folder')
train_data, rest_data = data.split(0.8)
validation_data, test_data = rest_data.split(0.5)
model = image_classifier.create(train_data, validation_data=validation_data)
loss, accuracy = model.evaluate(test_data)
config = QuantizationConfig.for_float16()
model.export(export_dir='path-to-save-model', quantization_config=config,export_format=ExportFormat.TFLITE)
model.export(export_dir='path-to-save-label', quantization_config=config,export_format=ExportFormat.LABEL)

**Comment by aboubacryba** [comment]
> use this code to train you custom model
> 
> import os
> 
> import numpy as np
> 
> import tensorflow as tf assert tf.**version**.startswith('2')
> 
> from tflite_model_maker import model_spec from tflite_model_maker import image_classifier from tflite_model_maker.config import ExportFormat from tflite_model_maker.config import QuantizationConfig from tflite_model_maker.image_classifier import DataLoader
> 
> import matplotlib.pyplot as plt
> 
> #to unzip a rar !unzip path-of-zip-file -d path-to-save-extract-file
> 
> data = DataLoader.from_folder('path-of-custom-folder') train_data, rest_data = data.split(0.8) validation_data, test_data = rest_data.split(0.5) model = image_classifier.create(train_data, validation_data=validation_data) loss, accuracy = model.evaluate(test_data) config = QuantizationConfig.for_float16() model.export(export_dir='path-to-save-model', quantization_config=config,export_format=ExportFormat.TFLITE) model.export(export_dir='path-to-save-label', quantization_config=config,export_format=ExportFormat.LABEL)

You just saved my life. Thank You !!!!!
